### ISSUE TITLE
Word boundaries for Alt-B, Alt-F, Ctrl-W, Alt-Left, Alt-Right... changed with 3.4.0 update

### ISSUE THREAD

**Issue by bo0ts** [fish-shell/fish-shell#8790]
Since updating to fish version 3.4.0 the behaviour of Alt-B, Alt-F, and Cmd-W has changed. For example

    $ asdf_adsf # pressing Cmd-W used to delete _asdf, now deletes asdf_asdf

I can't find documentation for this change and would like to be able to revert back to the old behaviour. This seems related to the discussion in issue #384

**Comment by faho** [comment]
We're reasonably happy with how these things work now and won't be changing them back.

E.g. ctrl-w is "backward-kill-path-component", it's supposed to remove one path component, and now it does.

Sorry if that's not to your liking.

**Comment by bo0ts** [comment]
I see why the change to `Ctrl-W` is sensible, but why change all other movement keys as well? This behaviour should depend on emacs or vi mode. The new behaviour is not consistent with either mode.

**Comment by bo0ts** [comment]
@faho Unfortunatel Alt-B and Ctrl-W now have a different notion of "word":

    $ foo/bar # Ctrl-W
    $ foo/
    $ foo/bar # cursor at the end, Alt-B now jumps to beginning

**Comment by faho** [comment]
They do, because "ctrl-w" has no notion of a "word". It has a notion of a "path-component", it acts on path components.

**Comment by bo0ts** [comment]
@faho A lot of thought went into the previous behaviour and it was consistent as can be seen from #384. This change throws out all the old considerations, introduces inconsistencies and in compatabilities.

Could you at least consider making this configurable?

**Comment by faho** [comment]
>A lot of thought went into the previous behaviour and it was consistent

And a lot of thought went into the current behavior, including complaints about the old being inconsistent.

>Could you at least consider making this configurable?

We don't make things like this configurable.

I am afraid the answer is quite simply: No. I now consider this matter closed. Sorry!

**Comment by krobelus** [comment]
> Since updating to fish version 3.4.0 the behaviour of Alt-B, Alt-F, and Cmd-W has changed.
>
>    $ asdf_adsf # pressing Cmd-W used to delete _asdf, now deletes asdf_asdf

Same thing happened on 3.3.1.

I think the behavior you want is `bind \cw backward-kill-word`

Your observed behavior change might stem from e50d10439 (Remove duplicate \cw binding in default key bindings, 2015-11-08).
That commit removed the `backward-kill-word` C-w binding; now only `backward-kill-path-component` remains.

I don't think Alt-B or Alt-F have changed much (ever)?

Regarding configurability: maybe it's worth adding commands like `forward-regex` (C-s), `backward-kill-regex` etc.
They can be used to implement many possible motions.
When used without a regex argument, they can prompt for one interactively. I think that'd be nice to navigate long commandlines. Today I just use my editor via A-e, that's good enough too.

**Comment by bo0ts** [comment]
@krobelus Thank you for the helpful comment. @faho Turns out this is already perfectly customisable.

```
bind \eb backward-word
bind \ef forward-word
bind \cw backward-kill-word
```

**Comment by amw** [comment]
I'm sorry, I am confused about what is the expected behavior now and I can't find any mention of this change in the release notes.

From experience I can say I've been driven up the wall these past two days as it seems Alt stopped working for me in the most basic scenario like this:

1. I type `cd`
2. Fish suggests completion based on history like `Documents/Computer/Configuration/` (displays it in muted color in front of my cursor)
3. I press Alt+Right Arrow

What used to happen is that it would complete the first path component and leave the other two as suggestion.

Now it just completes the whole path – just as if I pressed Right Arrow alone without holding Alt key. I can't choose a different directory in `Documents/` unless I manually type in the whole directory name.

What am I missing? What else am I supposed to use instead of Alt+Right Arrow?

**Comment by krobelus** [comment]
> What used to happen is that it would complete the first path component and leave the other two as suggestion.

Yes, that's the expected behavior.
You have either remapped Alt+RightArrow, or your terminal sends a funny key sequence when you press that key

Here's how you can check what key sequence is sent, and its current binding (mine is the default)

	$ fish_key_reader
	Press a key:
	bind \e\[1\;3C 'do something'
	$ bind \e\[1\;3C
	bind --preset \e\[1\;3C nextd-or-forward-word

**Comment by amw** [comment]
I use `Terminal.app` on macOS and I don't believe I've changed any mapping from the default. Attached screenshot of "Keyboard" section from profile preferences.

```
fish_key_reader
Press a key:
bind \ef 'do something'
```

<img width="667" alt="Screen Shot 2022-03-22 at 21 41 50" src="https://user-images.githubusercontent.com/29783/159572290-e1b45338-b2e3-4725-9924-491dad0e1b1f.png">

**Comment by faho** [comment]
@amw Your terminal sends alt+f instead of alt+right. This is indistinguishable to fish.

You want to check that "Use Option as Meta key" to unbreak it.

**Comment by amw** [comment]
Unfortunately that did not help. For what it's worth I've never played with that Meta key option before and Alt+Right seemed to work. What changed in 3.4?

**Comment by faho** [comment]
Try `fish_key_reader` again.

Nothing springs to mind that changed in 3.4, and sending \ef (which *is* alt+f) is wrong and will do something different from sending something recognizable as alt+right.

**Comment by faho** [comment]
Nope, found it - in #2330 we worked around Terminal.app's awful key sequences, so this now does "nextd-or-forward-word" like the alt+right binding is supposed to.

And in #8538 we reimplemented that in C++, and that has

https://github.com/fish-shell/fish-shell/blob/1f997de8bf9a5402f79d2d39c7500b2e9b379bcb/src/reader.cpp#L3552-L3565

See how that last line sets the movement style for "forward_word" but not for "nextd_or_forward_word"? Yeah.

**Comment by faho** [comment]
Okay, should be fixed.

I'm going to cherry-pick that one for 3.4.1, which should be released soon.

**Comment by amw** [comment]
Setting `bind --preset \ef forward-word` did make Alt+Right as well as Alt+f do the right thing for me. Should I add that to my personal config or will it work out of the box after the patch?

**Comment by krobelus** [comment]
nice, thanks for the report & fix. I guess our nightlies haven't been tested much with Terminal.app

> will it work out of the box after the patch?

yes

**Comment by amw** [comment]
Thank you for quick support and your work on the project!

Me being "driven up the wall" by this for the last two days is just an indication of how important Fish has been in my life in the last 6+ years I've been using it. It's part of my brain now.

**Comment by zanchey** [comment]
> I guess our nightlies haven't been tested much with Terminal.app

I use it but I guess I don't use Alt-F much

**Comment by faho** [comment]
Note: The original request turned out to be a misunderstanding about what the ctrl-w binding does (this was changed in 2012, not recently).

What was fixed was a regression in alt-left/alt-right.

So we're playing a bit fast and loose with this issue. That probably can't be helped - opening a separate issue just for that won't help much.

**Comment by rajashekar007** [comment]
Sorry to comment on the closed issue, but I'm encountering with fish_3.4.1:
fish: Unknown command: nextd-or-forward-word

**Comment by faho** [comment]
> I'm encountering an issue with fish_3.4.1

@rajashekar007 
You're not running fish 3.4.1. That's an old fish left over.

**Comment by 42xel** [comment]
@faho I have the same issue on 3.7.1. Also, I was running 3.3.1 later this morning which I installed not six months ago. But that's just me not reading the manual and taking whatever version was on dnf without adding any repo.

Alt-B and Alt-F work fine, which really surprises me as I would have expected them to be binding for the same function.
On an other machine, everything works fine. The main difference is the good one is Ubuntu, the one with an error is Oracle linux server.

Don't try too hard if nothing comes to mind immediately, I'll do a proper bug report after some further digging. I just had to comment that first § itwas just too funny bringing up my frozen fish right after you scolded someone for months old leftovers.
